rfcs/rfc-0000-discriminators.md

```diff
@@ -0,0 +1,123 @@
+- This should curb any incentive to sell / steal accounts for "rare usernames".
+
+  Certain "rare" discriminators can also be prohibited.
+
+- Allow users who do not want to be easily discoverable to stay hidden.
+
+  Just a 4-digit discriminator provides 10,000 unique users.
+  Expanding on this, we could for example, let users hide discriminators in servers and similar privacy features.
+
+# Guide-level explanation
+
+Revolt would be switching to a new username system made up of three parts:
+
+```
+Hello, World!#123456
+^ your chosen name
+             ^ separator between chosen name and discriminator
+              ^ discriminator
+```
+
+Discriminators for new users are generated randomly on sign up while existing users will
+get a randomly generated discriminators upon the server software being upgraded.
+
+> Revolt treats usernames as case-insensitive so a full username such as "Jason#1234" is
+> effectively the same as "jasON#1234" but only "Jason#1234" will display in the UI.
+
+Users may:
+
+- Change the case of their username at any moment, such as from "Jason" to "jasON".
+
+  This will never affect their discriminator.
+
+- Change their username, but may have their discriminator regenerated if there is a conflict.
+
+  Such as if you are changing from "Jason#1234" to "Phil", but "Phil#1234" is taken.
+
+### User Opinion
+
+Prior discussions on the forums primarily seem to lean in favour of discriminators being
+implemented into Revolt, the internal team has not raised any complaints against it, and
+the primary community on Revolt ("Revolt Lounge") appears to be somewhat split on the
+issue.
+
+As a recent case study, Discord's change to unique usernames only has generated a lot of
+backlash and criticism from the community, it appears to be overwhemingly negative but
+these experiences are mostly coming from long-term users of Discord. The true opinion of
+the average user has not been gauged yet.
+
+### Usability
+
+This may make it more difficult to understand how to add users on Revolt, however this
+shouldn't be an issue as long as it's clearly communicated through the user interface.
+Revolt already allows a wide range of Unicode in usernames so we are definitely not
+making a significant UX hit towards the usability of the friend system.
+
+> There are further considerations listed at the end of this document which must be
+> addressed if this RFC were to be implemented.
+
+# Reference-level explanation
+
+WIP
+
+This is the technical section of the RFC, it should go over in detail:
+
+- Its interaction with other features
+- How this will be implemented
+- Corner or edge cases
+
+This section should reference the examples in the previous section and disect them in more detail.
+
+# Drawbacks
+
+WIP
+
+Why should this not be added.
+
+# Rationale and alternatives
+
+WIP
+
+- Why is this design the best
+- Are there alternative ways to solve this
```

Comment:
i'm personally not a huge fan of sharing "friend code" links or using stuff like QR codes. these sorts of things can potentially be used for phishing ops and other scams. sharing usernames directly is also much more intuitive, while using codes like this could feel somewhat clunky.

it almost certainly *would* fix many of the potential issues with discriminators, but i think connecting with others could also become more of a challenge.

Reply:
also, what if someone wants to have their revolt profile open for anyone to add? like on a website in their social links or whatever. could friend codes be made permanent and infinitely usable like server invites? and if so, couldn't we essentially find ourselves right back at square one with the issue of unique codes potentially being rare and valuable?

Reply:
> the issue i have with friend codes is the fact that they're cosmetically very hard to identify. a quasirandom string of numbers and letters gives no obvious clues on who the code belongs to, unlike a username+discriminator pair.

maybe in this case, you could append the user's display name to the invite link? since functionally the code is the only part that would really matter, you could just have the server ignore everything after the initial code is read.

almost certainly a REALLY hacky solution, but it'd probably work lol

Reply:
I think this has all been mentioned through the RFC.